### PR TITLE
Makefile work (version improvments, help target)

### PR DESCRIPTION
**Pull request by wadells**
In preparation for Robotest 2.0 (nothing fancy, it is a formality to acknowledge previous breaking changes) and to address #200 this changeset refactors the Robotest Makefile to have robust semver compliant versioning and docker tagging.  `make help` thrown in as a bonus.

I explored using the following solutions, in addition to the one proposed here:

 1. https://github.com/gravitational/version Test branch here: https://github.com/wadells/robotest/commit/97b20ed0d17d78e51d91de15753d3c0b38163481  I dismissed this because I found the linkflags passing of information and extra dependecy to be obtuse compared to a generated `version.go`.

2. https://github.com/mdomke/git-semver This repo has clearly grappled with the exact same problems that motivated my changes.  I still found that <40 lines of code in make/shell to be prefereable to a new tool in the toolchain that isn't used elsewhere in gravitational.

### Warning!
This will cause our semver version to regress from `1.0.2XX` to `1.0.1-dev.2.XX`  However, I plan to create a 2.0 alpha tag as soon as this is in.  As for 1.0.X: There have already been many breaking changes in 1.0.X patch releases, so the 1.0.X line of development is not properly semver compliant anyhow.  I do not plan to continue to support 1.0.X once 2.0.0 is out.

### Testing Done:
<details>
<summary><code>make help</code> testing</summary>
<pre>
walt@work:~/git/robotest$ make help
Usage: make [options] [target] ...

Options: run `make --help` for options

Targets:
  all             Clean and build.
  build           Compile go binaries.
  clean           Remove intermediate build artifacts & cache.
  containers      Build container images.
  help            Show this message.
  lint            Run static analysis against source code.
  publish         Publish container images to quay.io.
  test            Run unit tests.
  version         Show the robotest version.
</pre>
</details>

<details>
<summary><code>make version</code> testing</summary>
<pre>
walt@work:~/git/robotest$ make version
version metadata saved to version.go
Robotest Version: 1.0.1-dev.221+d40bd462
walt@work:~/git/robotest$ git tag 2.0.0-alpha
walt@work:~/git/robotest$ make version       
version metadata saved to version.go
Robotest Version: 2.0.0-alpha
walt@work:~/git/robotest$ git tag -d 2.0.0-alpha
Deleted tag '2.0.0-alpha' (was d40bd46)
walt@work:~/git/robotest$ git tag 2.0.0-alpha HEAD~5
walt@work:~/git/robotest$ make version
version metadata saved to version.go
Robotest Version: 2.0.0-alpha.5+d40bd462
walt@work:~/git/robotest$ git tag -d 2.0.0-alpha 
Deleted tag '2.0.0-alpha' (was a5daaa8)
</pre>
</details>

<details>
<summary><code>robotest-suite -version</code> testing</summary>
<pre>
walt@work:~/git/robotest$ make build
docker build --pull --tag robotest:buildbox \
        --build-arg UID=$(id -u) \
        --build-arg GID=$(id -g) \
        --build-arg GOLANGCI_LINT_VER=1.21.0 \
        docker/build
Sending build context to Docker daemon  3.072kB
Step 1/12 : FROM quay.io/gravitational/debian-venti:go1.12.9-stretch
go1.12.9-stretch: Pulling from gravitational/debian-venti
Digest: sha256:db307e15860db30beb731c5da4b0e3a4b329963cec62579755743f6f631aaec9
Status: Image is up to date for quay.io/gravitational/debian-venti:go1.12.9-stretch
 ---> 9575e05c4a00
Step 2/12 : ARG UID
 ---> Using cache
 ---> 4eb7355a8b88
Step 3/12 : ARG GID
 ---> Using cache
 ---> ad2a81726301
Step 4/12 : ARG GOLANGCI_LINT_VER
 ---> Using cache
 ---> 14684ffe8de0
Step 5/12 : ENV GOPACKAGESPRINTGOLISTERRORS=true
 ---> Using cache
 ---> 417ce996fb46
Step 6/12 : RUN groupadd builder --gid=$GID -o;     useradd builder --uid=$UID --gid=$GID --create-home --shell=/bin/bash;
 ---> Using cache
 ---> e52c6e240224
Step 7/12 : RUN (mkdir -p /go/src/github.com/gravitational/robotest && chown -R builder /go ${GOPATH})
 ---> Using cache
 ---> fe0d4311a124
Step 8/12 : RUN (mkdir -p /go/bin)
 ---> Using cache
 ---> 1df70b697a68
Step 9/12 : ENV LANGUAGE="en_US.UTF-8"     LANG="en_US.UTF-8"     LC_ALL="en_US.UTF-8"     LC_CTYPE="en_US.UTF-8"     GOPATH="/gopath"     PATH="$PATH:/opt/go/bin:/go/bin"
 ---> Using cache
 ---> ca2bfcf41f1c
Step 10/12 : RUN (wget -q https://github.com/golangci/golangci-lint/releases/download/v$GOLANGCI_LINT_VER/golangci-lint-$GOLANGCI_LINT_VER-linux-amd64.tar.gz &&     tar -xvf golangci-lint-$GOLANGCI_LINT_VER-linux-amd64.tar.gz -C /go/bin          golangci-lint-$GOLANGCI_LINT_VER-linux-amd64/golangci-lint --strip-components=1 &&   rm golangci-lint-$GOLANGCI_LINT_VER-linux-amd64.tar.gz)
 ---> Using cache
 ---> 1ec25b825a1d
Step 11/12 : WORKDIR /gopath/src/github.com/gravitational/robotest
 ---> Using cache
 ---> 8abadc810849
Step 12/12 : VOLUME ["/gopath/src/github.com/gravitational/robotest"]
 ---> Using cache
 ---> 5542b701b221
Successfully built 5542b701b221
Successfully tagged robotest:buildbox
mkdir -p build
docker run --rm=true -u $(id -u):$(id -g) -v /home/walt/git/robotest:/go/src/github.com/gravitational/robotest -v /home/walt/git/robotest/build:/go/src/github.com/gravitational/robotest/build -w /go/src/github.com/gravitational/robotest robotest:buildbox \
        dumb-init make -j e2e suite
version metadata saved to version.go
go version go1.12.9 linux/amd64
go version go1.12.9 linux/amd64
cd /go/src/github.com/gravitational/robotest && \
        GO111MODULE=on go test -mod=vendor -c -i ./suite -o build/robotest-suite
cd /go/src/github.com/gravitational/robotest && \
        GO111MODULE=on go test -mod=vendor -c -i ./e2e -o build/robotest-e2e
</pre>
<pre>
walt@work:~/git/robotest$ ./build/robotest-suite -version
Version:        1.0.1-dev.221+d40bd462
Git Commit:     d40bd462955892a129a013beb531d9d772bce927
</pre>
</details>

Quay.io publishing testing can be seen via the `1.0.1-dev.220-e42cd6ff` tag at https://quay.io/repository/gravitational/robotest-suite?tab=tags

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,22 +1,39 @@
 TARGETS := e2e suite
+VERSION ?= $(shell ./version.sh)
+GIT_COMMIT := $(shell git rev-parse HEAD)
+VERSIONFILE := version.go
+# docker doesn't allow "+" in image tags: https://github.com/docker/distribution/issues/1201
+DOCKER_VERSION ?= $(subst +,-,$(VERSION))
+TAG ?= latest
```
Comment (wadells) [inline Makefile:7]: In order to discourage use of named tags, I considered removing this variable & related infrastructure.  I believe you use this regularly @a-palchikov (particularly the `dima` tag) which is why I chose to leave it in place presently.

File: Makefile [hunk 2]
```diff
@@ -28,42 +45,66 @@ buildbox:
...
 lint: buildbox
 	docker run $(DOCKERFLAGS) \
 		--env="GO111MODULE=off" \
 		$(BUILDBOX) dumb-init golangci-lint run \
 		--skip-dirs=vendor \
 		--timeout=2m
+
+.PHONY: version
+version: ## Show the robotest version.
+version: $(VERSIONFILE)
```
Comment (a-palchikov) [inline Makefile:80]: I'm not sure investing in Makefile templating is worth it. We've been gradually migrating to [mage](https://github.com/magefile/mage) - some examples are [satellite](https://github.com/gravitational/satellite/blob/master/build.go), [planet](https://github.com/gravitational/planet/blob/master/build.go) and [wormhole](https://github.com/gravitational/wormhole/blob/master/build.go).

Comment (wadells) [reply Makefile:80]: I wasn't particularly fond of this templating either.  Its something I copied from https://github.com/gravitational/teleport/blob/8c75759b98fbf71f2932c6f605c7d8feb4e50dfd/version.mk as the least complex way of dynamically generating versions that I've found.

I'll take a look at mage.  I suspect that migration would considerably increase the scope of this PR though, so I may work on it as a separate work item.

**Comment by UnderGreen** [comment]
I agree with Dima. If we can afford to invest time in migration to Mage that would be awesome

File: Makefile (hunk 2 above)
Comment (knisbet) [reply Makefile:80]: You don't need to do a complete rewrite in Mage, new targets can be setup in mage / go, but existing stuff doesn't necessarily have to change right of way. And a make target can just call into mage as well as a stop gap.

Comment (wadells) [reply Makefile:80]: I took an hour yesterday to POC a target or two with mage.  I estimate maybe two full days of my time to migrate over to it (one for conversion, a 2nd for testing, jenkins cleanup & maybe PR builds #171).  As such I've filed https://github.com/gravitational/robotest/issues/207 for mage work.

> You don't need to do a complete rewrite in Mage, new targets can be setup in mage / go, but existing stuff doesn't necessarily have to change right of way.

Given the amount of work, I'd prefer not to leave things in a partially migrated state.  I will make the time for a full conversion before 2.0 (which I'm aiming for at the end of the month).

The remaining question is: Should we take the work in this PR, or scrap this until mage arrives?  I'm inclined to take it as the 80 lines here:

1) are largely consistent with how our other projects manage versions
2) won't make the mage migration significantly harder
3) provide some immediate value (specifically, it allows me to easily figure out which git commit a robotest container in the wild corresponds to)

**Comment by wadells** [comment]
> @wadells This has been sitting for a bit, is this good to merge?

Probably good to merge.  I hadn't clicked the button because I'd been waiting on a response to the `mage` thread here:

https://github.com/gravitational/robotest/pull/206#discussion_r410398331

I've have some half finished work porting the build to mage that will deprecate/replace the work in this PR, but I don't know when I'll find the priority to return to it.

**Pull request merged**